awx_collection/plugins/module_utils/tower_api.py

```diff
@@ -84,7 +84,10 @@ def __init__(self, argument_spec, **kwargs):
         if self.params.get('validate_certs') is not None:
             self.verify_ssl = self.params.get('validate_certs')
         if self.params.get('tower_oauthtoken'):
-            self.oauth_token = self.params.get('tower_oauthtoken')
+            if type(self.params.get('tower_oauthtoken')) is dict:
+                self.oauth_token = self.params.get('tower_oauthtoken')['token']
+            elif 'token' in self.params.get('tower_oauthtoken'):
+                self.oauth_token = self.params.get('tower_oauthtoken')
```

Comment:
After this line, the code should validate type of `self.oauth_token`, because it's no longer automatic. Here, only a string type should be acceptable.

Reply:
Not sure how you got the code like this, maybe an old commit? It looks like this:
```
        if self.params.get('tower_oauthtoken'):
            if type(self.params.get('tower_oauthtoken')) is dict:
                self.oauth_token = self.params.get('tower_oauthtoken')['token']
            elif 'token' in self.params.get('tower_oauthtoken'):
                self.oauth_token = self.params.get('tower_oauthtoken')
```

Reply:
We could I guess try and raise an exception if we fall down into an "else" condition here. LMK if you want me to do that.

Reply:
yeah, I'm also not really quite sure what's going on here, and if you actually intended this diff.

Backing up, what you're doing is turning off the validation for the token parameter. That's fine is documented, and we give clean error messages when it doesn't fit the expected pattern.

In several ways, it seems like this still isn't finished, or at least the validation isn't.

What are our expectations?

 - If it is a dict, then it needs to have the key "token", it can have other keys, but this is a minimum condition. Ansible handles `KeyError`s poorly, so if it's a dict, but that dict doesn't have the token key, this should fail with a message to document the failure of expectations.
 - If it's a string, then it should be used as-is
 - If it's a list, or any other type, then it should fail, repeating or referring to the basic documentation of the token parameter

The current code has a few head-scratchers. where is it handed as a string? `elif 'token' in self.params.get('tower_oauthtoken'):` is assuming a dict, but the prior conditional is also assuming a dict.

Reply:
i think that's suppsed to be `elif not 'token' in self.params.get('tower_oauthtoken'):` which gives us the old string behavior, my testing shows that the old string behavior is currently broken, im pretty sure that's why

Reply:
This has been updated, take a look and let me know what you think.

Reply:
👍